risc0/zkvm/src/receipt.rs

```diff
@@ -206,7 +211,10 @@ impl SegmentReceipts {
         final_receipt.verify_with_context(ctx)?;
         let metadata = final_receipt.get_metadata()?;
         log::debug!("final: {metadata:#?}");
-        if prev_image_id != metadata.pre.digest() {
+        // If there's more than one segment, the last segment could be the fault checker
+        if prev_image_id != metadata.pre.digest()
+            && !(receipts.len() > 0 && metadata.pre.digest() == FAULT_CHECKER_ID.into())
+        {
```

Comment:
I'm not very happy about this part of the code... I wish that we could set the exit_code from the circuit to be FAULT rather than SYSTEM_SPLIT but I don't really see any other way except changing the circuit... I wonder if there's anything actually wrong with this logic though: if we see a SYSTEM_SPLIT and a segment with FAULT_CHECKER_ID, is it safe to assume that the previous segment faulted??

Reply:
I don't know how we'd make the circuit do this. It seems like the Executor is the one detecting the condition.